---
title: "Introduction to xCell 2.0"
author: "Almog Angel"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{Introduction to xCell2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

`xCell2` is an R package for performing cell type enrichment analysis of bulk gene expression data. It builds upon the methodology of the original `xCell` package, providing improved algorithms and enhanced performance. More importantly, xCell2 is generic, namely you can use any reference (including scRNA-Seq) to train an xCell2 reference object for the analysis.

This vignette provides an overview of the package's functionality, including how to prepare input data, use the main functions, and perform basic cell type enrichment analysis.

# Installation

(Soon) To install the `xCell2` package from Bioconductor, use:

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("xCell2")
```

To install the `xCell2` package source code, use:

```{r, eval = FALSE}
devtools::install_github("AlmogAngel/xCell2")
```

# Generating custom xCell2 reference object using xCell2Train

This section contain two parts.
In the first part we will prepare all the require inputs for `xCell2Train`.
In the second part we learn use the `xCell2Train` to generate a new xCell2 reference object.

## Preparing required inputs

In this example we will use a sub-set of 1561 bulk RNA-seq samples from sorted cell populations from the Database of Immune Cell Expression (DICE) by Schmiedel B et al. (2018).

You can access the full reference data using the `celldex` package.
However, here we will use a sub-set of the DICE reference just for learning purpose.

```{r, eval = TRUE}
# Access the full DICE referece:
# dice <- celldex::DatabaseImmuneCellExpressionData() 

# Load a small sub-set of the DICE referece:
data(dice_demo_ref, package = "xCell2")
```

The `xCell2Train` function requires a reference cell type gene expression matrix and a data.frame with different labels the correspond to the samples/cells in the reference.

*Reference gene expression matrix:*
- The reference gene expression matrix can be generated from any sequencing platform - Micro-array, bulk RNA-Seq and single-cell RNA-Seq.
- xCell2 rely on enrichment scores of gene-sets, therefore reference can be in raw counts or any normalization and xCell2 can handle data in either linear or logarithmic space.

*Labels data frame:*
- In addition to the reference, xCell2 requires a corresponding data frame with different labels for the samples/cells in the reference.
- The rows in the data frame should correspond to columns in the ref.
- The data frame must have four columns:
    1. "ont": the cell type ontology as a character (i.e., "CL:0000545" or NA if there is no ontology).
    2. "label": the cell type name as a character (i.e., "T-helper 1 cell").
    3. "sample": the cell type sample/cell that match the column name in ref.
    4. "dataset": sample's source dataset or subject (can be the same for all samples if no such information).


```{r, eval = TRUE}
# Extract reference matrix
dice_ref <- as.matrix(dice_demo_ref@assays@data$logcounts)
colnames(dice_ref) <- make.unique(colnames(dice_ref)) # Make samples samples unique

# Extract reference metadata
dice_labels <- as.data.frame(dice_demo_ref@colData)

# Prepare labels data frame
dice_labels$ont <- NA
dice_labels$sample <- colnames(dice_ref)
dice_labels$dataset <- "DICE"
```

### Assing cell type ontology

*You can skip the following step if:*
  *a) You don't want to use ontology to avoid cell type dependencies (not recommended)*
  *b) You are sure that there are no cell type dependencies in your reference*

Find the cell type ontology here: https://www.ebi.ac.uk/ols4/ontologies/cl
For example, search "T cells, CD4+, memory" and look for the best match ("CD4-positive, alpha-beta memory T cell" - "CL:0000897")
```{r, eval = TRUE}
dice_labels[dice_labels$label == "B cells",]$ont <- "CL:0000236"
dice_labels[dice_labels$label == "Monocytes",]$ont <- "CL:0000576"
dice_labels[dice_labels$label == "NK cells",]$ont <- "CL:0000623"
dice_labels[dice_labels$label == "T cells, CD8+",]$ont <- "CL:0000625"
dice_labels[dice_labels$label == "T cells, CD4+",]$ont <- "CL:0000624"
dice_labels[dice_labels$label == "T cells, CD4+, memory",]$ont <- "CL:0000897"
```


You can manually check cell type dependencies using the xCell2GetLineage function:
```{r, eval = FALSE}
xCell2::xCell2GetLineage(labels = dice_labels, outFile = "demo_dice_dep.tsv") 
```
Open `demo_dice_dep.tsv` and check if the lineage assigned to each cell type makes sense.
Note that "T cells, CD4+, memory" assigned as a descendant of "T cells, CD4+"

## Generate custom xCell2 reference object

Now when dice_ref and dice_labels are ready, we can generate a new xCell2 object!

Simply run the following command:
```{r, eval = TRUE}
set.seed(123) # (optional) As generating pseudo-bulk samples from scRNA-Seq reference based on random sampling of cells
DICE_demo.xCell2Ref <- xCell2::xCell2Train(ref = dice_ref, labels = dice_labels, refType = "rnaseq")
```

*xCell2Train Parameters:*

*ref* A reference gene expression matrix (genes in rows, samples/cells in columns)
*mix* A bulk mixture of gene expression data (genes in rows, samples in columns) (optional) - if you provide the mixture, xCell2Train will learn cell type signatures using only the shared genes between the reference and mixture and might results a reference that is better suit to your mixture.
*labels* A data frame in which the rows correspond to samples in the ref. The data frame must have four columns:
    1. "ont": the cell type ontology as a character (i.e., "CL:0000545" or NA if there is no ontology).
    2. "label": the cell type name as a character (i.e., "T-helper 1 cell").
    3. "sample": the cell type sample/cell that match the column name in ref.
    4. "dataset": sample's source dataset or subject (can be the same for all samples if no such information).
*refType* The reference gene expression data type: "rnaseq" for bulk RNA-Seq, "array" for micro-array, or "sc" for scRNA-Seq.
*minPbCells* For scRNA-Seq reference only - minimum number of cells in the pseudo-bulk (optional, default: 30).
*minPbSamples* For scRNA-Seq reference only - minimum number of pseudo-bulk samples (optional, default: 10).
*minScGenes* For scRNA-Seq reference only - minimum number of genes for pseudo-bulk samples (default: 10000).
*useOntology* A Boolean for considering cell type dependencies by using ontological integration (default: TRUE).
*lineageFile* Path to the cell type lineage file generated with `xCell2GetLineage` function and reviewed manually (optional) - In our example: "demo_dice_dep.tsv".
*numThreads* Number of threads for parallel processing (default: 1).
*humanToMouse* A Boolean for converting human genes to mouse genes (default: FALSE).
*topSpillValue* Maximum spillover compensation correction value (default: 0.5).
*returnSignatures* A Boolean to return just the signatures (default: FALSE).
*returnAnalysis* A Boolean to return the xCell2Analysis results (do not return reference object) (default: FALSE).
*useSpillover* A Boolean to use spillover correction in xCell2Analysis (returnAnalysis must be TRUE) (default: TRUE).
*spilloverAlpha* A numeric for spillover alpha value in xCell2Analysis (returnAnalysis must be TRUE) (default: 0.5).

### Share your custom xCell2 reference object with the scientific community!

Once generated a new custom xCell2 reference, you can share this object with the so other can use it.

1. Save your reference object
```{r, eval = FALSE}
save(DICE_demo.xCell2Ref, file = "DICE_demo.xCell2Ref.rda")
```

2. Go to: https://github.com/dviraran/xCell2refs/

3. Go to the "references" directory

4. Click "Add file" > "Upload files" and upload your reference

5. Go back to  https://github.com/dviraran/xCell2refs/ and to "README.md"

6. Click on the pencil icon ("Edit this file") on the top right corner

7. Modify this table with information about your reference

# Running xCell2Analysis

To run cell type enrichment analysis using xCell 2.0 you need an xCell2 reference object

You can generate a custom reference (see the section above) or load "ready to use" references make by the scientific community.
Information about the references avialable for xCell 2.0 can be found here: https://github.com/dviraran/xCell2refs/

In this example we will load a demo xCell2 reference object DICE_demo.xCell2Ref (made above).
```{r, eval = TRUE}
?xCell2::DICE_demo.xCell2Ref
data(DICE_demo.xCell2Ref, package = "xCell2")
```

Now you can run `xCell2Analysis`:
```{r, eval = TRUE}
data(mix_demo, package = "xCell2")
xcell2_res <- xCell2::xCell2Analysis(mix = mix_demo, xcell2object = DICE_demo.xCell2Ref)
```

*xCell2Analysis Parameters:*

*mix* A bulk mixture of gene expression data (genes in rows, samples in columns).
NOTE: mix and xcell2object must have the same genes annotations.
*xcell2object* An S4 object of class `xCell2Object`.
*minSharedGenes* Minimum fraction of shared genes required between the mix and the reference (default: 0.9).
*rawScores* Boolean to indicate whether to return raw enrichment scores (default: FALSE).
*spillover* Boolean - use spillover correction on the transformed enrichment scores? (default: TRUE).
*spilloverAlpha* A numeric for spillover alpha value (default: 0.5).
*numThreads* Number of threads for parallel processing (default: 1).


```{r, eval = TRUE}
sessionInfo()
```