#' A subset SummarizedExperiment object of the DICE reference for xCell 2.0 vignette
#'
#' This is a demo reference object to learn how to use 'xCell2Train'
#'
#' @source Schmiedel B et al. (2018)

"dice_demo_ref"

#' An xCell2 object trained from the Tumor Microenvironment Compendium reference
#'
#' Platform: RNA-Seq
#' Organism: Human
#' Tissue: Tumor
#'
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference curated by Zaitsev A (2022) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::TMECompendium.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"TMECompendium.xCell2Ref"

#' An xCell2 object trained from the Tabula Muris reference
#'
#' Platform: single-cell RNA-Seq
#' Organism: Mouse
#' Tissue: Immune/Blood
#'
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference generated by The Tabula Muris Consortium (2018) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::TabulaMurisBlood.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"TabulaMurisBlood.xCell2Ref"

#' An xCell2 object trained from the Tabula Sapiens reference
#'
#' Platform: single-cell RNA-Seq
#' Organism: Human
#' Tissue: Immune/Blood
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference generated by The Tabula Sapiens Consortium (2022) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::TabulaSapiensBlood.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"TabulaSapiensBlood.xCell2Ref"

#' An xCell2 object trained from the MouseRNAseqData reference
#'
#' Platform: RNA-Seq
#' Organism: Mouse
#' Tissue: General
#'
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference generated by Benayoun B (2019) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::MouseRNAseqData.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"MouseRNAseqData.xCell2Ref"

#' An xCell2 object trained from the PanCancer reference
#'
#' Platform: single-cell RNA-Seq
#' Organism: Human
#' Tissue: Tumor
#'
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference generated by Nofech-Mozes I (2023) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::PanCancer.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"PanCancer.xCell2Ref"

#' An xCell2 object trained from the LM22 reference
#'
#' Platform: Micro-array
#' Organism: Human
#' Tissue: General
#'
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference generated by Newman AM (2015) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::LM22.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"LM22.xCell2Ref"

#' An xCell2 object trained from the Immune Compendium reference
#'
#' Platform: RNA-Seq
#' Organism: Human
#' Tissue: Immune/Blood
#'
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference curated by Zaitsev A (2022) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::ImmuneCompendium.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"ImmuneCompendium.xCell2Ref"

#' An xCell2 object trained from the Immunologic Genome Project reference
#'
#' Platform: Micro-array
#' Organism: Mouse
#' Tissue: Immune/Blood
#'
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference generated by The Immunological Genome Project Consortium (2008), curated by Aran D (2019) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::ImmGenData.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"ImmGenData.xCell2Ref"

#' An xCell2 object trained from the Blueprint and ENCODE Projects reference
#'
#' Platform: RNA-Seq
#' Organism: Human
#' Tissue: General
#'
#' You can use this reference object as in input for `xCell2Analysis` or create a new reference object using `xCell2Train`.
#'
#' @format An xCell2Object (S4) that contain:
#' \describe{
#'   \item{\code{params}}{A tibble of parameters for linear transformation.}
#'   \item{\code{signatures}}{A list containing cell-type signatures.}
#'   \item{\code{dependencies}}{An list specifying cell types dependencies.}
#'   \item{\code{spill_mat}}{A matrix representing spillover correction factors.}
#'   \item{\code{genes_used}}{A character vector of genes used to generate the reference.}
#' }
#' @source Reference generated by Martens JHA and Stunnenberg HG (2013), The ENCODE Project Consortium (2012), curated by Aran D (2019) and trained by Angel A, et al (2024).
#'
#' @examples
#' xCell2Ref <- xCell2::BlueprintEncode.xCell2Ref
#' results.xcell2 <- xCell2Analysis(mix, xcell2object = xCell2Ref)
"BlueprintEncode.xCell2Ref"
